bin/list-all

```diff
@@ -16,8 +12,8 @@ sort_versions() {
 
 versions=$(
   eval "$cmd $releases_path" |
-    grep -o '"name": "v.*' |
-    sed 's/"name": "v\(.*\)",/\1/' |
+    grep -o '### v\?\([0-9]\{1,2\}[.]\)\{2\}[0-9]\{1,2\}\([-]rc[0-9]\)\?' |
+    sed 's/### v\?//' |
```

Comment:
As much as it's a tad ugly, we should probably use the [suggested regular expression](https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string) to check a SemVer string, rather than reinventing the wheel.

Also, [`grep` supports `-E`](https://linux.die.net/man/1/grep) (short for `--extended-regexp`) and [`sed` supports `-r`](https://linux.die.net/man/1/sed#:~:text=%2Dr%2C%20%2D%2Dregexp%2Dextended) (short for `--regexp-extended`), which configure the tools to use regular expressions (instead of their own (imo, confusing and error prone) pattern languages). I believe this makes the commands more readable and maintainable.

```suggestion
    grep -oE '### v?((0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?)' | # suggested RegEx for SemVer https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
    sed -E 's/### v?//' |
```

Reply:
I thank you for the review and suggestion. But this regex is only catching 92 occurrences.

Reply:
Good catch!

The regex worked for me on macOS 12, but I tried on Ubuntu (via docker), and also got only 92 matches. I modified the regex to use `[0-9]` in place of `\d` and that seemed to work in both environments. I guess `\d` isn't supported by Ubuntu's `grep`.

Will you try it?

```suggestion
    grep -oE '### v?((0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)(?:-((?:0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?)' | # suggested RegEx for SemVer (with minor modification for better grep compatibility) https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
    sed -E 's/### v?//' |
```

Reply:
It worked here! But the `-rc1~3` versions are not being caught, I don't know about macOS because unfortunately I can't test it, and I even tried to see why here, but I'm not good with Regex. On the other hand, I have a question/suggestion, in relation to getting the alphanumerics, why not use `[a-Z]` instead of `[a-zA-Z-]`?

Reply:
No problem! Happy to help you learn.

I think `grep` on Ubuntu didn't like the non-capturing groups (the groups that started with `?:`). Since we care about the full match (not the individual captures), I simply removed the `?:`s (making them capturing groups), and it worked on Ubuntu in Docker.

```suggestion
    grep -oE '### v?((0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)(-((0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*)(\.(0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*))*))?(\+([0-9a-zA-Z-]+(\.[0-9a-zA-Z-]+)*))?)' | # suggested RegEx for SemVer (with minor modification for better grep compatibility) https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
    sed -E 's/### v?//' |
```

Hopefully, this is the one!

Reply:
Wow, it worked!

Of course it worked, you are very good! I thank you for the teachings.